design/src/rfcs/rfc0026_custom_validation_exceptions.md

```diff
@@ -0,0 +1,671 @@
+  response body for validation failures would be able to be migrated to use a
+  smithy-rs server SDK.
+
+Currently, it is a _strict requirement_ that constrained operations's errors
+contain `smithy.framework#ValidationException`, and its error messages are not
+configurable.
+
+### Solution proposal
+
+It is worth noting that the _vast_ majority of smithy-rs server SDK users are
+operating within a backend service, and that the default behavior prescribed by
+Smithy's protocol tests is satisfactory for their use cases. Therefore, it is
+reasonable to expect that a solution to lift the aforementioned restriction be
+_opt-in_.
+
+We will introduce a `codegenConfig.disableDefaultValidation` flag in
+`smithy-build.json` with the following semantics:
+
+* If `disableDefaultValidation` is set to `false`, we will continue operating
+  as we currently do: all constrained operations _must_ have
+  `smithy.framework#ValidationException` attached.
+* If `disableDefaultValidation` is set to `true`, any constrained operation
+  that does not have `smithy.framework#ValidationException` attached will be
+  code generated so as to force the user to provide a [`core::ops::Fn`] value
+  that maps from the constrained operation's `ConstraintViolationExceptions`
+  container newtype to any modeled error shape attached to the operation.
+
+For example, an operation like:
+
+```smithy
+operation MyOperation {
+    input: MyOperationInput,
+    output: MyOperationOutput,
+    errors: [ValidationCustomError, OperationError1, ... OperationErrorN]
+}
+
+@error("client")
+structure ValidationCustomError {
+    @required
+    validationErrorMessage: String
+}
+```
+
+in a model generated with `codegenConfig.disableDefaultValidation` set to
+`true` will be implemented in user code as:
+
+```rust
+use my_server_sdk::{input, output, error}
+
+fn map_operation_validation_failure_to_error(errors: input::my_operation_input::ConstraintViolationExceptions) -> error::OperationError {
+    let mut validation_error_message: String::new();
+
+    for e in errors.into_iter() {
+        match e {
+            // ...
+        }
+    }
+
+    // We can return _any_ error attached to the operation.
+    error::OperationError::ValidationCustomError(
+        error::ValidationCustomError {
+            validation_error_message
+        }
+    )
+}
+
+async fn operation_handler(input: input::MyOperationInput) -> Result<output::MyOperationOutput, error::MyOperationError> {
+    // ...
+}
+```
+
+Since the operation does not have `smithy.framework#ValidationException`
+attached, when registering the handler in the service builder, the user must
+provide the mapping to an `error::MyOperationError` variant when constraint
+trait enforcement fails.
+
+```rust
+let my_service = MyService::builder_without_plugins()
+    .my_operation(operation_handler)
+    .on_my_operation_validation_failure(map_operation_validation_failure_to_error)
```

Comment:
Should this API be on the `Operation` struct?

For example,

```rust
let op = MyOperation::from_handler(handler).custom_validation(closure);
```

- This is more inline with how we apply specific layers to operations (`Operation::layer`) - a decision was made to not put more methods on the service builder [here](https://github.com/awslabs/smithy-rs/blob/main/design/src/rfcs/rfc0020_service_builder.md#approach-d-add-more-methods-to-the-service-builder).
- More closely resembles the typescript API, [related](https://github.com/awslabs/smithy-rs/pull/2053/files#diff-26f3492c91346ff7493644f733cd3391b418f88fb0d2099c7305b399e695e75bR377).

Reply:
Came here to say basically the same thing 😄.